### PR TITLE
[JBPM-9197] Allow the introduction of emails in Tasks notifications

### PR DESCRIPTION
**Pull request by elguardian**
allow arbitrary emails notifications through parameter toemails

### CODE REVIEW

**Comment by elguardian** [comment]
depends on https://github.com/kiegroup/droolsjbpm-knowledge/pull/445

**Comment by elguardian** [comment]
Jenkins retest this please

**Comment by elguardian** [comment]
Jenkins retest this please

**Comment by elguardian** [comment]
Jenkins retest this please

File: jbpm-human-task/jbpm-human-task-jpa/src/main/java/org/jbpm/services/task/persistence/JPATaskPersistenceContext.java [hunk 1]
```diff
@@ -225,12 +227,49 @@ public User updateUser(User user) {
...
+        check();
+        return this.em.merge(email);
+    }
+
+    @Override
+    public Email removeEmail(Email email) {
+        check();
+        em.remove(email);
+        return email;
+    }
```
Comment (gmunozfe) [inline jbpm-human-task/jbpm-human-task-jpa/src/main/java/org/jbpm/services/task/persistence/JPATaskPersistenceContext.java:271]: I cannot see where these methods are invoked and if they are really needed.
For persist, at the end, _persistIfNotExist_ calls to _tpc.persistOrgEntity(entity);_
Could you clarify if they are needed for persistence? 
I cannot see an analogous scenario to _add-group-command_ for email.

File: jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/commands/UserGroupCallbackTaskCommand.java [hunk 2]
```diff
@@ -391,9 +411,10 @@ protected void doCallbackOperationForTaskDeadlines(Deadlines deadlines, TaskCont
...
                                                 doCallbackUserOperation(recipient.getId(), context);
-                                            }
-                                            if(recipient instanceof Group) {
+                                            } else if (recipient instanceof Group) {
                                                 doCallbackGroupOperation(recipient.getId(), context);
+                                            } else if (recipient instanceof Email) {
+                                                doCallbackEmailOperation(recipient.getId(), context);
                                             }
                                         }
                                     }
```
Comment (gmunozfe) [inline jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/commands/UserGroupCallbackTaskCommand.java:420]: This block (L407-420) is repeated also for End Deadlines. Perhaps, now it's a good opportunity to refactor it and extract it into a private method for enhancing maintainability.

File: jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/commands/UserGroupCallbackTaskCommand.java [hunk 3]
```diff
@@ -98,6 +100,14 @@ protected boolean doCallbackUserOperation(String userId, TaskContext context) {
 
     }
     
+    private boolean doCallbackEmailOperation(String emailId, TaskContext context) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/commands/UserGroupCallbackTaskCommand.java:103]: ok

File: jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/commands/UserGroupCallbackTaskCommand.java [hunk 4]
```diff
@@ -144,12 +154,22 @@ protected User addUserFromCallbackOperation(String userId, TaskContext context)
...
     }
     
+    protected Email addEmailFromCallbackOperation(String emailId, TaskContext context) {
+        Email email = context.getPersistenceContext().findEmail(emailId);
+        boolean emailExists = email != null;
+        if (!StringUtils.isEmpty(emailId) && !emailExists) {
+            persistIfNotExists(TaskModelProvider.getFactory().newEmail(emailId), context);
+        }
+
+        return email;
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/commands/UserGroupCallbackTaskCommand.java:164]: ok

File: jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/commands/UserGroupCallbackTaskCommand.java (hunk 2 above)
Comment (elguardian) [reply jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/commands/UserGroupCallbackTaskCommand.java:420]: done

File: jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/deadlines/notifications/impl/email/EmailNotificationListener.java [hunk 5]
```diff
@@ -78,43 +79,37 @@ public void onNotification(NotificationEvent event, UserInfo userInfo) {
...
     
                     Message msg = new MimeMessage(mailSession);
                     Set<String> toAddresses = new HashSet<String>();
-                    for (User user : entry.getValue()) {
+                    for (OrganizationalEntity user : entry.getValue()) {
     
-                        String emailAddress = userInfo.getEmailForEntity(user);
-                        if (emailAddress != null) {                        	
+                        String emailAddress = getEmailFromOrganizationEntity(userInfo, user);
+                        if (emailAddress != null) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/deadlines/notifications/impl/email/EmailNotificationListener.java:112]: done.

File: jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/deadlines/notifications/impl/email/EmailNotificationListener.java [hunk 6]
```diff
@@ -218,28 +221,33 @@ protected URL getAttachemntURL(String attachment) throws MalformedURLException {
...
 	        }
     	}
     }
 
-    protected void buildMapByLanguage(Map<String, List<User>> map, User user, UserInfo userInfo) {
-        String language = userInfo.getLanguageForEntity(user);
-        List<User> list = map.get(language);
+    protected void buildMapByLanguage(Map<String, List<OrganizationalEntity>> map, OrganizationalEntity user, UserInfo userInfo) {
+        String language = Optional.ofNullable(userInfo.getLanguageForEntity(user)).orElse("en-UK");
+        List<OrganizationalEntity> list = map.get(language);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/deadlines/notifications/impl/email/EmailNotificationListener.java:240]: done

File: jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/commands/UserGroupCallbackTaskCommand.java (hunk 2 above)
Comment (elguardian) [reply jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/commands/UserGroupCallbackTaskCommand.java:420]: ok

File: jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/identity/MvelUserGroupCallbackImpl.java [hunk 7]
```diff
@@ -20,21 +20,27 @@
...
 
 import org.jbpm.services.task.utils.MVELUtils;
+import org.kie.api.task.model.Email;
 import org.kie.api.task.model.Group;
 import org.kie.api.task.model.User;
 import org.kie.internal.task.api.UserGroupCallback;
 
 public class MvelUserGroupCallbackImpl implements UserGroupCallback {
 
+    private Set<Email> emails;
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/identity/MvelUserGroupCallbackImpl.java:37]: removed

File: jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/impl/util/HumanTaskHandlerHelper.java [hunk 8]
```diff
@@ -98,13 +100,17 @@ public static Deadlines setDeadlines(Map<String, Object> parameters, List<Organi
...
         
         for (String component : allComponents) {
-	        String[] mainComponents = component.split(ELEMENT_SEPARATOR);
-	        
-	        if (mainComponents!= null && mainComponents.length == 2) {
-	            String actionComponent = mainComponents[0].substring(1, mainComponents[0].length()-1);
-	            String expireComponents = mainComponents[1].substring(1, mainComponents[1].length()-1);
-	 
-	            String[] expireElements = expireComponents.split(ATTRIBUTES_ELEMENTS_SEPARATOR);
+            Pattern pattern = Pattern.compile("\\[(.*?)\\]@\\[(.*?)\\]");
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/impl/util/HumanTaskHandlerHelper.java:103]: done

File: jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/impl/util/HumanTaskHandlerHelper.java [hunk 9]
```diff
@@ -193,27 +199,30 @@ protected static Deadline getNewDeadline(String expiresAt, Date deadlineDate, St
...
 
-				for (String id : groupRecipientsIds) {
-					Group group = TaskModelProvider.getFactory().newGroup();
-                	((InternalOrganizationalEntity) group).setId(id.trim());
-					notificationRecipients.add(group);
-				}
-			}
+            String emailRecipients = parameters.get("toemails");
+            if (emailRecipients != null && emailRecipients.trim().length() > 0) {
+                String[] emailRecipientsIds = emailRecipients.split(ATTRIBUTES_ELEMENTS_SEPARATOR);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/impl/util/HumanTaskHandlerHelper.java:221]: for consistency I let this.

File: jbpm-human-task/jbpm-human-task-jpa/src/main/java/org/jbpm/services/task/persistence/JPATaskPersistenceContext.java [hunk 10]
```diff
@@ -225,12 +227,49 @@ public User updateUser(User user) {
...
+    public Email persistEmail(Email email) {
+        check();
+        try {
+            this.em.persist(email);
+            if (this.pessimisticLocking) {
+                this.em.flush();
+                return this.em.find(EmailImpl.class, email.getId(), LockModeType.PESSIMISTIC_WRITE);
+            }
+        } catch (EntityExistsException e) {
+            throw new RuntimeException("Email already exists with " + email + " id, please check that there is no group and user with same id");
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-human-task/jbpm-human-task-jpa/src/main/java/org/jbpm/services/task/persistence/JPATaskPersistenceContext.java:255]: done

File: jbpm-human-task/jbpm-human-task-jpa/src/main/java/org/jbpm/services/task/persistence/JPATaskPersistenceContext.java (hunk 1 above)
Comment (elguardian) [reply jbpm-human-task/jbpm-human-task-jpa/src/main/java/org/jbpm/services/task/persistence/JPATaskPersistenceContext.java:271]: for consistency leting this here.

**Comment by elguardian** [comment]
Regarding validation or stunner should be in stunner pr not here.

ok for others scenarios.

**Comment by elguardian** [comment]
> Overall looks good, but I think following scenario is not completely covered:
> 
> Currently, `getLanguageForEntity(OrganizationalEntity entity)` works in the Test with the `DefaulUserInfo` implementation, because it returns null, and the Optional uses `.orElse("en-UK");`
> 
> For userinfo.properties, it could be defined with the email as key (user) like:
> _[tony@domain.com](mailto:tony@domain.com)=[tony@domain.com](mailto:tony@domain.com):fr-FR:tony_
> 
> However, with other implementation of UserInfo, like `DBUserInfoImpl`, the customized _db.query.lang_ might be something like this:
> 
> _select lang from Users where userId = ? **or email = ?**_
> 
> In that case, `DBUserInfoImpl.getLanguageForEntity` should control that query is defined with 2 parameters and bind also the second one with the same _entity.getId()_.
> 
> ```
> if (ps.getParameterMetaData().getParameterCount() == 2) {
>   ps.setString(2, entity.getId());
> }
> ```
> 
> WDYT?

that is implementation specific. I would not change as the id is still what it means.

**Comment by gmunozfe** [comment]
> > Overall looks good, but I think following scenario is not completely covered:
> > Currently, `getLanguageForEntity(OrganizationalEntity entity)` works in the Test with the `DefaulUserInfo` implementation, because it returns null, and the Optional uses `.orElse("en-UK");`
> > For userinfo.properties, it could be defined with the email as key (user) like:
> > _[tony@domain.com](mailto:tony@domain.com)=[tony@domain.com](mailto:tony@domain.com):fr-FR:tony_
> > However, with other implementation of UserInfo, like `DBUserInfoImpl`, the customized _db.query.lang_ might be something like this:
> > _select lang from Users where userId = ? **or email = ?**_
> > In that case, `DBUserInfoImpl.getLanguageForEntity` should control that query is defined with 2 parameters and bind also the second one with the same _entity.getId()_.
> > ```
> > if (ps.getParameterMetaData().getParameterCount() == 2) {
> >   ps.setString(2, entity.getId());
> > }
> > ```
> > 
> > 
> > WDYT?
> 
> that is implementation specific. I would not change as the id is still what it means.

Thanks for your clarification, so in these cases, email should be also populated as userId (better to be renamed at custom tables as entityId)

**Pull request merged**